### PR TITLE
Add index to improve performance of the `/timestamp_to_event` endpoint used for jumping to a specific date in the timeline of a room.

### PR DESCRIPTION
**Pull request by reivilibre**
Follows: #9445, #14215 

Base: `develop` 

Original commit schedule, with full messages:

<ol>
<li>

Add index to `events` table to help with 'jump to date' \
Before

```sql
synapse=> explain             SELECT event_id FROM events
            LEFT JOIN rejections USING (event_id)
            WHERE
                room_id = '!...:matrix.org'
                AND origin_server_ts <= 1620172800000
                /**
                 * Make sure the event isn't an `outlier` because we have no way
                 * to later check whether it's next to a gap. `outliers` do not
                 * have entries in the `event_edges`, `event_forward_extremeties`,
                 * and `event_backward_extremities` tables to check against
                 * (used by `is_event_next_to_backward_gap` and `is_event_next_to_forward_gap`).
                 */
                AND NOT outlier
                /* Make sure event is not rejected */
                AND rejections.event_id IS NULL
            /**
             * First sort by the message timestamp. If the message timestamps are the
             * same, we want the message that logically comes "next" (before/after
             * the given timestamp) based on the DAG and its topological order (`depth`).
             * Finally, we can tie-break based on when it was received on the server
             * (`stream_ordering`).
             */
            ORDER BY origin_server_ts DESC, depth DESC, stream_ordering DESC
            LIMIT 1;
                                                    QUERY PLAN
------------------------------------------------------------------------------------------------------------------
 Limit  (cost=1075.38..2148.31 rows=1 width=66)
   ->  Incremental Sort  (cost=1075.38..650197.88 rows=605 width=66)
         Sort Key: events.origin_server_ts DESC, events.depth DESC, events.stream_ordering DESC
         Presorted Key: events.origin_server_ts
         ->  Nested Loop Anti Join  (cost=0.71..650170.66 rows=605 width=66)
               ->  Index Scan Backward using events_ts on events  (cost=0.43..649835.91 rows=605 width=66)
                     Index Cond: (origin_server_ts <= '1620172800000'::bigint)
                     Filter: ((NOT outlier) AND (room_id = '!...:matrix.org'::text))
               ->  Index Only Scan using rejections_event_id_key on rejections  (cost=0.28..0.55 rows=1 width=41)
                     Index Cond: (event_id = events.event_id)
```

Index

```sql
synapse=> create index rei_jtd_idx ON events(room_id, origin_server_ts) WHERE not outlier;
CREATE INDEX
synapse=> explain             SELECT event_id FROM events
            LEFT JOIN rejections USING (event_id)
            WHERE
                room_id = '!...:matrix.org'
                AND origin_server_ts <= 1620172800000
                /**
                 * Make sure the event isn't an `outlier` because we have no way
                 * to later check whether it's next to a gap. `outliers` do not
                 * have entries in the `event_edges`, `event_forward_extremeties`,
                 * and `event_backward_extremities` tables to check against
                 * (used by `is_event_next_to_backward_gap` and `is_event_next_to_forward_gap`).
                 */
                AND NOT outlier
                /* Make sure event is not rejected */
                AND rejections.event_id IS NULL
            /**
             * First sort by the message timestamp. If the message timestamps are the
             * same, we want the message that logically comes "next" (before/after
             * the given timestamp) based on the DAG and its topological order (`depth`).
             * Finally, we can tie-break based on when it was received on the server
             * (`stream_ordering`).
             */
            ORDER BY origin_server_ts DESC, depth DESC, stream_ordering DESC
            LIMIT 1;
                                                               QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=5.44..10.08 rows=1 width=66)
   ->  Incremental Sort  (cost=5.44..2819.10 rows=607 width=66)
         Sort Key: events.origin_server_ts DESC, events.depth DESC, events.stream_ordering DESC
         Presorted Key: events.origin_server_ts
         ->  Nested Loop Anti Join  (cost=0.83..2791.78 rows=607 width=66)
               ->  Index Scan Backward using rei_jtd_idx on events  (cost=0.56..2456.44 rows=607 width=66)
                     Index Cond: ((room_id = '!...:matrix.org'::text) AND (origin_server_ts <= '1620172800000'::bigint))
               ->  Index Only Scan using rejections_event_id_key on rejections  (cost=0.28..0.55 rows=1 width=41)
                     Index Cond: (event_id = events.event_id)
```

</li>
</ol>

### CODE REVIEW

**Review by squahtx — APPROVED** [review]
Looks good!

File: synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+INSERT INTO background_updates (ordering, update_name, progress_json) VALUES
+  (7324, 'events_jump_to_date_index', '{}');
```
Comment (H-Shay) [inline synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql:17]: For my own curiosity, what does ordering refer to here and how did you derive the value 7324 for it?

Comment (DMRobertson) [reply synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql:17]: My entirely home-grown speculative understanding:

There can sometimes be a causal dependency between background updates. E.g. if update A adds a new column and then update B adds an index to that column, B must happen after A.

Synapse's background update machinery ensures(?) that updates with smaller ordering happen before updates with larger ordering. (I'm guessing that somewhere we `SELECT * FROM background_updates ORDER BY ordering ASC LIMIT 1;` or similar to pick the upgrade to run next.)

As for the value: `7324` is the concatenation of `73` and `24`, taken from the file path; `73` is the schema compat version and `24` is the migration version within schema compat `73.`

Comment (H-Shay) [reply synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql:17]: Thank you!

Comment (clokep) [reply synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql:17]: Note you can also add an explicit dependency via the `depends_on` column

Comment (DMRobertson) [reply synapse/storage/schema/main/delta/73/24_events_jump_to_date_index.sql:17]: https://github.com/matrix-org/synapse/blob/373c485d8c7f39206bac60c6ef313b4a1978bbc0/synapse/storage/background_updates.py#L375-L411 has the machinery, I think

**Pull request merged**